### PR TITLE
[JENKINS-63711] Cleanup GitHub App Credential cache folders 

### PR DESCRIPTION
**Pull request by bitwiseman**
# Description
Before GitHub App Credentials, cache folders not being cleaned up was okay.  It was sometimes even an advantage since cached values that were extremely stable didn't have to be reloaded even across connection instances.  However, with  GitHub App Credentials the password (token) is regenerated on a regular basis, resulting in new connections with new caches.  

This change cleans up cache folders when GitHub App connections are no longer used. Behavior for non-GitHubAppCredential connections is unchanged to preserve the advantages of the existing behavior for standard credentials.

See [JENKINS-63711](https://issues.jenkins-ci.org/browse/JENKINS-63711) for further information. 

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [ ] Automated tests have been added to exercise the changes
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verify that the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

# Documentation changes
- [ ] Link to jenkins.io PR, or an explanation for why no doc changes are needed

# Users/aliases to notify

### CODE REVIEW

**Review by MarkEWaite — APPROVED** [review]
Didn't perform interactive testing, but the code looks reasonable to me.

**Comment by bitwiseman** [comment]
I'm not sure how to go about testing this - the behavior was (obviously) untested before this change.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:603]: No need to use a boxed `Long` here. It adds overhead, and makes it possible for someone to accidentally introduce NPEs in the future since the current code assumes it is non-null.

```suggestion
        private long lastUsed = System.currentTimeMillis();
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:653]: With the above suggestion:

```suggestion
                long lastUse = record.lastUsed;
```

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:659]: Should we catch `IOException` here so that even if there is a problem with one connection we still try to clean up the rest? (We could still throw it at the end of the method; just add suppressed exceptions if more than one cache has an issue)

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java
Comment (dwnusbaum) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:537]: I was concerned about deadlocks at first because we are introducing a new lock, but it looks like this block never does any other synchronization, so I think it should be ok.

**Review by dwnusbaum — APPROVED** [review]
Looks fine in general; added some suggestions.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java
Comment (car-roll) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:659]: +1 for cleanup

**Review by car-roll — APPROVED** [review]

**Comment by drahamim** [comment]
Is there a reason this isn't merged yet? I got hit with the cache issue and it dead stopped work till I cleaned up the directory. 
It caused a storage issue where our disk was maxed out on Inodes before it even reached 50% capacity.

**Comment by MarkEWaite** [comment]
@drahamim one of the great persuaders for an open source maintainer is when a user says:

> I installed the [pre-release build](https://ci.jenkins.io/job/Plugins/job/github-branch-source-plugin/job/PR-341/6/artifact/org/jenkins-ci/plugins/github-branch-source/2.9.2-rc1275.3848d510f09d/github-branch-source-2.9.2-rc1275.3848d510f09d.hpi).  It resolves my issue and is working well.

Have you installed the pre-release build?  Is it working well for you?

**Comment by drahamim** [comment]
@MarkEWaite Valid point. I will see if I can do that ( and look up how to do that).

**Comment by MarkEWaite** [comment]
That would be great @drahamim .  The pre-release build link in my message is the file that you would download to your local computer then upload into Jenkins from "Manage Plugins" .  The upload is available in the "Advanced" tab.

**Comment by drahamim** [comment]
@MarkEWaite  How easy is it to switch back to mainline after utilizing pre-release? 
I got approval  but have that one concern.

**Comment by MarkEWaite** [comment]
Usually a switch back to the preceding release is as easy as pressing the plugin manager's "Downgrade" button and then restart Jenkins (assuming you use the plugin manager's facility to upload the pre-release plugin to your Jenkins installation).

**Comment by kad-kostej** [comment]
Hi, I have a question. Suppose this issue will be solved and released. Will it also clean the old cache contents ? Some time ago, i had to clean the cache manually for just 1 Jenkins instance: it took 20 hours. If the plugin does that in the near future, will it be done in the background or will it "hang" my Jenkins instance ? I need to know that because it will have impact on our production facilities (40 Jenkins instances).

File: src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java
Comment (bitwiseman) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/Connector.java:692]: Added this in the last change.  It is likely the existing users of github app credentials will have a large number of untracked cache folders.  This will attempt to clean them out.

**Review by dwnusbaum — APPROVED** [review]

**Comment by bitwiseman** [comment]
@dwnusbaum @kad-kostej 
I started to have it do cleanup but your example made me think it was better not to.

**Comment by timja** [comment]
Possibly caused https://issues.jenkins.io/browse/JENKINS-64595

**Pull request merged**
